File: source/smallsmt-openpnp-server/commsockets.cpp
#include "commsockets.h"

int CommSockets::initSockets()
{

    ourSocket = new QUdpSocket(this);
    ourSocket->bind(QHostAddress::LocalHost, SOCKET_NR_OUR);

    theirSocket = new QUdpSocket(this);

    connect(ourSocket, SIGNAL(readyRead()),this, SLOT(readPendingDatagrams()));

    return 0;
}


void CommSockets::sendToOpenPnp(QString msg)
{
  theirSocket->writeDatagram(msg.toUtf8(), QHostAddress::LocalHost,SOCKET_NR_THEIR);
}


void CommSockets::readPendingDatagrams()
{
    while (ourSocket->hasPendingDatagrams())
    {
        QNetworkDatagram datagram = ourSocket->receiveDatagram();
        processTheDatagram(datagram);
    }
}

void CommSockets::processTheDatagram(QNetworkDatagram datagram)
{
    emit recevieLog(QString(datagram.data()));
    sendToOpenPnp("ok");
}


CommSockets::CommSockets()
{
    initSockets();
}

File: source/smallsmt_openpnp_server/smallsmt_openpnp_server/commsockets.py
from PyQt5.QtNetwork import QUdpSocket,QHostAddress,QNetworkDatagram
from PyQt5.QtCore import QObject,pyqtSignal,pyqtSlot
import sys

class CommSockets(QObject):
    SOCKET_NR_OUR   = 9070
    SOCKET_NR_THEIR = 9072

    getFromOpenPnp = pyqtSignal(str)

    def __init__(self):
        super(CommSockets, self).__init__()

        self.ourSocket = QUdpSocket(self)
        self.ourSocket.bind(QHostAddress.LocalHost, self.SOCKET_NR_OUR);

        self.theirSocket = QUdpSocket(self)
        self.ourSocket.readyRead.connect(self.readPendingDatagrams)

    def sendToOpenPnp(self,message):
        self.theirSocket.writeDatagram(message.encode('utf-8'),QHostAddress.LocalHost, self.SOCKET_NR_THEIR)

    def processDatagram(self,datagram):
        self.getFromOpenPnp.emit(str(datagram.data(), encoding='utf-8'))

    @pyqtSlot()
    def readPendingDatagrams(self):

        while(self.ourSocket.hasPendingDatagrams()):
            datagram = self.ourSocket.receiveDatagram()
            self.processDatagram(datagram)

